runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -69,41 +71,49 @@ public <R, T extends ICorfuSMR<T>> R access(ICorfuSMRProxyInternal<T> proxy,
                                                 ICorfuSMRAccess<R, T> accessFunction,
                                                 Object[] conflictObject) {
         log.debug("Access[{},{}] conflictObj={}", this, proxy, conflictObject);
+
         // First, we add this access to the read set
         addToReadSet(proxy, conflictObject);
 
-        // Next, we sync the object, which will bring the object
-        // to the correct version, reflecting any optimistic
-        // updates.
         // Get snapshot timestamp in advance so it is not performed under the VLO lock
         long ts = getSnapshotTimestamp().getSequence();
-        return proxy
-                .getUnderlyingObject()
-                .access(o -> {
-                            WriteSetSMRStream stream = o.getOptimisticStreamUnsafe();
-
-                            // Obtain the stream position as when transaction context last
-                            // remembered it.
-                            long streamReadPosition = knownStreamsPosition.getOrDefault(proxy.getStreamID(), ts);
-
-                            return (
-                                    (stream == null || stream.isStreamCurrentContextThreadCurrentContext())
-                                    && (stream != null && getWriteSetEntrySize(proxy.getStreamID()) == stream.pos() + 1
-                                       || (getWriteSetEntrySize(proxy.getStreamID()) == 0 /* No updates. */
-                                          && o.getVersionUnsafe() == streamReadPosition) /* Match timestamp. */
-                                    )
-                            );
-                        },
-                        o -> {
-                            // inside syncObjectUnsafe, depending on the object
-                            // version, we may need to undo or redo
-                            // committed changes, or apply forward committed changes.
-                            syncWithRetryUnsafe(o, getSnapshotTimestamp(), proxy,
-                                    o::setUncommittedChanges);
-                        },
-                        accessFunction::access,
-                        version -> updateKnownStreamPosition(proxy.getStreamID(), version)
-        );
+
+        // TODO: better proxy type check, or refactor to avoid.
+        if (proxy instanceof CorfuCompileProxy) {
```

Comment:
I think this abstraction needs to be cleaned up. The problem here is that `ICorfuSMRProxyInternal` is not generic enough to capture capture both usages. I suggest removing `CorfuCompileProxy::underlyingObject` and pushing the `VersionLockedObject` logic further down.

Reply:
This would also mean making `AbstractTransactionalContext` more generic by removing `syncWithRetryUnsafe(VersionLockedObject`